Added a Virtual Pointer sample to the SDK

wyaneva opened the pull request:
This pull request contains:

* a new sample application - `example_vptr` which adds two matrices in SYCL using the Virtual Pointer interface
* a new template parameter for the type of buffer data
* test for default values of the template parameters in `utils/vptr/tests/basic.cc`

wyaneva commented:
While this sample compiles and runs, it does **not** produce correct results. I need to look into it closer.

I created the PR to facilitate discussion about the sample.

DuncanMcBain commented:
Sorry, I didn't see you were still updating this! I'll leave my comments for now, as they were mostly small things I thought I might fixup myself. But for now I'll step back! :)

wyaneva commented:
Hi, Duncan. Yes, there is still work I need to do on it. I will let you know when it is done :)

Ruyk commented:
Can you also fix the travis script to add the   COMPUTECPP_SDK_VIRTUAL_POINTER_PREFIX variable to the SDK?

DuncanMcBain commented:
As I mentioned in a previous comment, I'd prefer the SDK to build with as little user input as possible - to that end, since we know exactly where the header will be, I think we can just reference it directly in the CMake.

wyaneva commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,128 @@
+ * result in the third buffer. It then verifies the result of the kernel
+ * on the host by using a host accessor to gain access to the data. */
+int main() {
+  /* Destroying SYCL objects like queues and buffers blocks until all work
```
Yes, I copied code from another class - I will go over all comments in this one and tidy them up.

wyaneva commented:
Thanks for the comments :) I am still working on it (although a bit slow as I am in Bulgaria), will let you know when I am done.

jwlawson commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,129 @@
+     * circumstances, allocate space for it for the user (it will not
+     * allocate space when the map_allocator is used and an initial host
+     * pointer is provided, as this instructs the runtime to map the data
+     * into the host's memory). Since this code is attempting to access
```
Is there a sample using a map allocator? Should that be linked here?
It feels a bit strange to me that you are bringing up something that isn't used here and doesn't really have any direct relevance to this sample.

wyaneva replied:
Removed irrelevant comments. 0c2d66be8f9dee86bcaa729af98eca7ff31cbfa0

jwlawson commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,129 @@
+          range<1>{N * M}, [=](id<1> index) { B[index] = index[0] * 2014; });
+    });
+
+    /* This kernel will actually perform the computation C = A * B. Since
```
Should be `C = A + B`.

wyaneva replied:
Done. 0c2d66be8f9dee86bcaa729af98eca7ff31cbfa0

wyaneva commented:
Hi John, thanks for the comments. 
Regarding the buffer type - it does seem to be necessary and hopefully the use of default values would allow existing code using the vptr to remain unaffected.
I also think that changing the example to use the pointer arithmetics is a good idea and will look into doing that. Thanks for all the other comments too - it is still work in progress so I have some tidying up to do.
I am back in Edinburgh on Sept 4th and expect to address all of this then.

rodburns commented:
Is this ready to be merged into the master branch?

wyaneva commented:
Hi Rod, no, I am currently on holiday, I will address all the issues on Monday, Sept 4th.

wyaneva commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,128 @@
+ *
+ *  Codeplay's ComputeCpp SDK
+ *
+ *  example_sycl_application.cpp
```
Done.

wyaneva commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,128 @@
+ * result in the third buffer. It then verifies the result of the kernel
+ * on the host by using a host accessor to gain access to the data. */
+int main() {
+  /* Destroying SYCL objects like queues and buffers blocks until all work
```
Removed irrelevant comments. 0c2d66be8f9dee86bcaa729af98eca7ff31cbfa0

wyaneva commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,128 @@
+
+    /* Create device-only 2D buffers of floats for the matrices. */
+    /*
+    buffer<float, 2> a(range<2>{N, M});
```
Done. 73a94ea4b2cf0e9d030ed4539731f8a158f1a0a0

wyaneva commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,129 @@
+/***************************************************************************
+ *
+ *  Copyright (C) 2016 Codeplay Software Limited
```
Done. 0c2d66be8f9dee86bcaa729af98eca7ff31cbfa0

wyaneva commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,129 @@
+     * or on the host, the SYCL runtime would ensure that the data were
+     * copied between contexts etc. properly. */
+    myQueue.submit([&](handler& cgh) {
+      auto A = pMap.get_access<access::mode::write,
```
Done. 0c2d66be8f9dee86bcaa729af98eca7ff31cbfa0

wyaneva commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,129 @@
+    myQueue.submit([&](handler& cgh) {
+      auto A = pMap.get_access<access::mode::write,
+                               access::target::global_buffer, float>(a, cgh);
+      auto B = pMap.get_access<access::mode::write,
```
Done. 0c2d66be8f9dee86bcaa729af98eca7ff31cbfa0

wyaneva commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,128 @@
+     * to the device with no dependencies between each other. */
+    myQueue.submit([&](handler& cgh) {
+      auto B = pMap.get_access<access::mode::write>(b, cgh);
+      cgh.parallel_for<class init_b>(range<1>{N*M}, [=](id<1> index) {
```
Fixed. 74af77f73178937beb1f323e2a4587165f2cdd35

wyaneva commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,129 @@
+      auto A = pMap.get_access<access::mode::write,
+                               access::target::global_buffer, float>(a, cgh);
+      cgh.parallel_for<class init_a>(
+          range<1>(N * M), [=](id<1> index) { A[index] = index[0] * 2; });
```
Done. 74af77f73178937beb1f323e2a4587165f2cdd35

wyaneva commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,129 @@
+    /* This kernel enqueue will initialise buffer a. The accessor "A" has
+     * write access. */
+    myQueue.submit([&](handler& cgh) {
+      auto A = pMap.get_access<access::mode::write,
```
Done. 3c9a167feb140ea5913c6e7b39d29f0bffcaba5c

wyaneva commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,129 @@
+     * actually independent of each other. Therefore, they can be enqueued
+     * to the device with no dependencies between each other. */
+    myQueue.submit([&](handler& cgh) {
+      auto B = pMap.get_access<access::mode::write,
```
Done. 3c9a167feb140ea5913c6e7b39d29f0bffcaba5c

wyaneva commented on samples/example_vptr/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+set(SOURCE_NAME "example_vptr")
+
+set(CMAKE_CXX_STANDARD 11)
```
You are correct - removed the line: a79876d50b7a7cfec4d0314a2481a7a13f8d9b14

wyaneva commented on samples/example_vptr/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD_REQUIRED on)
+
+if(NOT COMPUTECPP_SDK_VIRTUAL_POINTER_PREFIX)
```
I understand - I think that keeping the samples easy to build is a good thing to aim for. I have now hard-coded `COMPUTECPP_SDK_VIRTUAL_POINTER_PREFIX` in the global file (`samples\CMakeLists.txt`) and include the directory there. a79876d50b7a7cfec4d0314a2481a7a13f8d9b14

Feel free to reorganise as you see fit :)

wyaneva commented:
Hi all, I am now done with this! It is ready for a final review.

In particular, since your last reviews I have changed the following:
1. Edited the source code comments, removing irrelevant ones.
2. Added pointer arithemtics in the indexing of the matrix on the host. @jwlawson, what do you think?
3. Tidied up the CMake files. @DuncanMcBain, is this what you had in mind?

@rodburns, @jwlawson and @Ruyk - do you think this is a useful example to demonstrate the use of virtual pointer?

Ruyk commented on utils/vptr/tests/basic.cc:
```diff
@@ -210,3 +210,30 @@ TEST(pointer_mapper, multiple_alloc_free) {
+    });
+
+    auto hostAcc = pMap.get_access<sycl_acc_rw, sycl_acc_host>(myPtr);
+    ASSERT_EQ(hostAcc[0], 1.0f);
```
can you wrap the hostAcc around a explicit scope? ({ .. }) that avoids hostAcc destructor called after the SYCLfree. It may work on the current computecpp implementation but there is no guarantee.

wyaneva replied:
Done. 05c45360a1f3058fac648fd8e2c9e7b1963c424b

Ruyk commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,115 @@
+
+    /* Similarly, this kernel will initialise the buffer pointed to by b. */
+    myQueue.submit([&](handler& cgh) {
+      auto B = pMap.get_access<access::mode::discard_write,
```
can you use different names altoghether? (accB or something). Uppercase/lowercase may be confusing to some people.

wyaneva replied:
Good point. Done. 05c45360a1f3058fac648fd8e2c9e7b1963c424b

DuncanMcBain commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,115 @@
+ *
+ *  Description:
+ *    Sample code that demonstrates the use of the virtual pointer intrface in
+ *SYCL on matrix addition.
```
spacing

wyaneva replied:
Done. 05c45360a1f3058fac648fd8e2c9e7b1963c424b

DuncanMcBain commented on utils/vptr/include/virtual_ptr.hpp:
```diff
@@ -247,11 +259,13 @@ class PointerMapper {
-            sycl_acc_target access_target = sycl_acc_target::global_buffer>
+  template <sycl_acc_mode access_mode = default_acc_mode,
+            sycl_acc_target access_target = default_acc_target,
+            typename buffer_data_type = buffer_data_type_t>
```
This is a larger interface change request, but if we moved the buffer_data_type to be the first template parameter, we could do stuff like:
```c++
auto p = static_cast<float *>(SYCLMalloc(/* some size*/));
... // in command group scope
auto a = get_access<float>(p, cgh);
```
That way we could default the mode and target most of the time, since it is likely the type that will most often need to change. That said if you don't think it's a good idea, let me know. It's an interface change for the sake of a little typing, so no big deal :)

keryell replied:
You can provide both versions to avoid breaking existing codes...

wyaneva replied:
I see what you mean, Duncan. To be honest I don't know how much existing code that might break is out there.

Ideally, I would like to dispense with needing to pass the data type at all. I have an idea and will try to do this in another MR.

keryell commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,115 @@
+     * the virtual pointer.*/
+    myQueue.submit([&](handler& cgh) {
+      auto A = pMap.get_access<access::mode::discard_write,
+                               access::target::global_buffer, float>(a, cgh);
```
I guess `, float` can be inferred here and is redundant in this example.
Could it be used then to do a `reinterpret_cast` when required?
Then you can remove all the `float` remaining in the example.

wyaneva replied:
Unfortunately, as the code is at the moment, we need to provide the data type, as we don not save it when using SYCLmalloc and it cannot be inferred. It would be great if we could infer it though. I will look into doing this in another MR.

keryell commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,115 @@
+    /* Allocate the matrices using SYCLmalloc. a, b and c are virtual pointers,
+     * pointing to device buffers.
+     */
+    float* a = static_cast<float*>(SYCLmalloc(N * M * sizeof(float), pMap));
```
Ahem... This looks like 40 year old C... :-)
What about:
```
auto a = SYCLmalloc<float>(N * M, pMap);
```
?

wyaneva replied:
:) 

I think your suggestion looks better, but I think the interface of `SYCLmalloc()` was deliberately chosen to match `std::malloc`.

What do others think?

Ruyk replied:
I actually like SYCLmalloc<float>, seems more straightforward.

Ruyk replied:
In any case, that should be a separate merge request so that this one can be closed :-)

keryell commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,115 @@
+    /* On the host, the result stored in the buffer of virtual pointer "c" are
+     * checked. The matrix is accessed row by row, using pointer arithmetics on
+     * the virtual pointer. */
+    float* c_row = c;
```
`auto c_row = c`

wyaneva replied:
Done. 444504bbbb3701a118247ecb499ad5843ef92d10

keryell commented on samples/example_vptr/example_vptr.cpp:
```diff
@@ -0,0 +1,115 @@
+    float* c_row = c;
+    for (size_t i = 0; i < N; i++) {
+      /* Get the number of elements by which the row is offset. */
+      auto row_offset = pMap.get_offset(c_row) / sizeof(float);
```
I guess we should provide an interface to return the index directly for the basic type, to avoid the division.

wyaneva replied:
Good idea. Done. 444504bbbb3701a118247ecb499ad5843ef92d10

wyaneva commented:
Hi all, if you are happy I think this PR can be merged now. I will make another one to polish type passing to the vptr functions.

Ruyk commented:
Looks good to me, can you sort out the conflicts? Thanks!
